src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferHashDataOverComputeHash.cs

```diff
@@ -0,0 +1,318 @@
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(StringRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            var hashAlgoBaseType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemSecurityCryptographyHashAlgorithm);
+            if (hashAlgoBaseType is null)
+            {
+                return;
+            }
+
+            var sha256Type = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemSecurityCryptographySHA256);
+            if (sha256Type is null)
+            {
+                return;
+            }
+
+            var byteType = context.Compilation.GetSpecialType(SpecialType.System_Byte);
+            if (byteType is null)
+            {
+                return;
+            }
+
+            var byteArrayParameter = ParameterInfo.GetParameterInfo(byteType, isArray: true, arrayRank: 1);
+
+            // method introduced in .NET 5.0
+            var hashDataMethodType = sha256Type.GetMembers(HashDataMethodName).OfType<IMethodSymbol>().GetFirstOrDefaultMemberWithParameterInfos(byteArrayParameter);
+            if (hashDataMethodType is null)
+            {
+                return;
+            }
+
+            var computeHashMethodBaseType = hashAlgoBaseType.GetMembers(ComputeHashMethodName).OfType<IMethodSymbol>().GetFirstOrDefaultMemberWithParameterInfos(byteArrayParameter);
+            if (computeHashMethodBaseType is null)
+            {
+                return;
+            }
+
+            context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+            return;
+
+            void OnOperationBlockStart(OperationBlockStartAnalysisContext context)
+            {
+                var computeHashVariableMap = PooledConcurrentDictionary<ILocalReferenceOperation, IInvocationOperation>.GetInstance();
+                var createdSymbolMap = PooledConcurrentDictionary<ILocalSymbol, DeclarationTuple>.GetInstance();
+                var localReferenceMap = PooledConcurrentDictionary<ILocalReferenceOperation, ILocalSymbol>.GetInstance();
+
+                context.RegisterOperationAction(CaptureHashLocalReferenceOperation, OperationKind.LocalReference);
+                context.RegisterOperationAction(CaptureCreateOrComputeHashInvocationOperation, OperationKind.Invocation);
+                context.RegisterOperationAction(CaptureHashObjectCreationOperation, OperationKind.ObjectCreation);
+                context.RegisterOperationBlockEndAction(OnOperationBlockEnd);
```

Comment:
So it seems like you have a moderately complex state based analysis per each operation block. Would it be possible to add a detailed comment with an example at the start of this method to explain the core logic. See here for a very nice example provided by @NewellClark: https://github.com/dotnet/roslyn-analyzers/blob/d464b5f144f57be88d6af32c779e5ce3949132fa/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryContainsMethods.cs#L79-L101

Reply:
I have added the comments